pkg/synthetictests/resource_growth.go

```diff
@@ -0,0 +1,105 @@
+	const testName = "[sig-trt] Secret count should not have grown significantly during upgrade"
+	return comparePostUpgradeResourceCountFromMetrics(testName, "secrets")
+
+}
+
+func testNoExcessiveConfigMapGrowthDuringUpgrade() []*junitapi.JUnitTestCase {
+	const testName = "[sig-trt] ConfigMap count should not have grown significantly during upgrade"
+	return comparePostUpgradeResourceCountFromMetrics(testName, "configmaps")
+
+}
+
+// comparePostUpgradeResourceCountFromMetrics tests that some counts for certain resources we're most interested
+// in potentially leaking do not increase substantially during upgrade.
+// This is in response to a bug discovered where operators were leaking Secrets and ultimately taking down clusters.
+// The counts have to be recorded before upgrade which is done in test/e2e/upgrade/upgrade.go, stored in a package
+// variable, and then read here in invariant. This is to work around the problems where our normal ginko tests are
+// all run in separate processes by themselves.
+//
+// Values for comparison are a guess at what would have caught the leak we saw. (growth of about 60% during upgrade)
+//
+// resource should be all lowercase and plural such as "secrets".
+func comparePostUpgradeResourceCountFromMetrics(testName, resource string) []*junitapi.JUnitTestCase {
+	oc := exutil.NewCLI("resource-growth-test")
+	ctx := context.Background()
+
+	preUpgradeCount := upgrade.PreUpgradeResourceCounts[resource]
+	e2e.Logf("found %d %s prior to upgrade", preUpgradeCount, resource)
+
+	cv, err := oc.AdminConfigClient().ConfigV1().ClusterVersions().Get(context.Background(), "version", metav1.GetOptions{})
+	if err != nil {
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+				FailureOutput: &junitapi.FailureOutput{
+					Output: "Error getting ClusterVersion: " + err.Error(),
+				},
+			},
+		}
+
+	}
+	upgradeCompletion := cv.Status.History[0].CompletionTime
+
+	// Use prometheus to get the resource count at the moment we recorded upgrade complete. We can't do this
+	// for the starting count as prometheus metrics seem to get wiped during the upgrade. We also don't want to
+	// just list the resources right now, as we don't know what other tests might have created since.
+	e2e.Logf("querying metrics at: %s", upgradeCompletion.Time.UTC().Format(time.RFC3339))
+	resourceCountPromQuery := fmt.Sprintf(`cluster:usage:resources:sum{resource="%s"}`, resource)
+	promResultsCompletion, err := helper.RunQueryAtTime(ctx, oc.NewPrometheusClient(ctx),
+		resourceCountPromQuery, upgradeCompletion.Time)
+	if err != nil {
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+				FailureOutput: &junitapi.FailureOutput{
+					Output: "Error getting resource count from Prometheus at upgrade completion time: " + err.Error(),
+				},
+			},
+		}
+
+	}
+	e2e.Logf("got %d metrics after upgrade", len(promResultsCompletion.Data.Result))
+	completedCount := float64(promResultsCompletion.Data.Result[0].Value)
```

Comment:
Are we guaranteed to have a `Result[0]`? Or above with the upgrade `History[0`]? A bounds check might be better in case something goes wrong, otherwise it'll just panic and not return any test result.

Reply:
Added